Document updated anchor calculations in .NET 8.0 (#7985)
dotnet/winforms: docs/design/anchor-layout-changes-in-net80.md

this.button7.Anchor = ((System.Windows.Forms.AnchorStyles)((System.Windows.Forms.AnchorStyles.Left | System.Windows.Forms.AnchorStyles.Right)));

// The following line triggers layout again to compute anchors with the default button size with a new location but still without a parent control.
this.button7.Location = new System.Drawing.Point(9, 47);

// The following line triggers layout once again to compute anchors with the button's new size but still without a parent control.
this.button7.Size = new System.Drawing.Size(134, 23);

// The following line triggers layout to compute anchors with button's current size and the default size for the parent control.
this.Controls.Add(button7);

// The following line triggers layout to compute anchors with the button's current size and a new size for the parent control.
// These sizes still may be changed for the current DPI depending on application's DPI mode.
this.Size = new System.Drawing.Size(828, 146);

// The following line triggers layout to compute anchors with the button's current size and new size for the parent control.

> // The current DPI  still not applied.

Reviewer: ```suggestion
// The current DPI is still not applied.
```

this.ResumeLayout(false)
```
The above snippet does not represent the complete set of instances where anchor computations are unnecessary and may hold invalid anchor values. It gets even more complicated when nested UserControls are involved.

## Known issues

Customer reported issues can be largely grouped as follows:

### Missing controls

Calculation of control's anchors with default sizes may result in negative anchors and, thus, result in an invalid location for the anchored controls.

![MissingControls](../images/AnchorLayoutKnownIssue_MissingControl.png)

### Overlapped Controls

Related to the "Missing controls" issue above, if the parent control is scaled to match the current monitor's DPI, and the control's anchor calculations happen out of sync with this scaling, controls may overlap:

![OverlappedControls](../images/AnchorLayoutKnownIssue_OverlappedControl.png)

## Anchor Calculations

> The following image illustrates anchors calculation with respect to a control's parent's display rectangle. The "red" rectangle is the parent's display rectangle, and the "green" rectangle is the anchored control's (button) bounds.

Reviewer 2: You have updated the image, but the text still refers to "green" and "red".

- `Left` arrow indicates the X coordinate of the button location with respect to the parent's display rectangle.
- `Top` arrow indicates the Y coordinate of the button location with respect to the parent's display rectangle.
- `Right` arrow indicates the distance from right edge of the parent's rectangle where button is placed.
- `Right` arrow indicates the distance from the bottom edge of the parent's rectangle where the button is placed.

    ![AnchorCalculations](../images/AnchorCalculations.png)

In the above image the boundary marked in "orange" is a container hosting the control (boundary marked with "blue"). The control can define its anchor property to explicitly tell the layout engine how its bounds change relative to its parent's bounds. For example: 
```CS
this.button14.Anchor = (System.Windows.Forms.AnchorStyles.Bottom
 | System.Windows.Forms.AnchorStyles.Left
 | System.Windows.Forms.AnchorStyles.Top
 | System.Windows.Forms.AnchorStyles.Right );
```

> When the control's `Anchor` property is set, the anchors (that is, left, top, right, bottom values) are computed and stored in an internal struct `AnchorInfo`. The only time the anchor values can be negative is when the control is placed/position outside/overlapped with the hosting controls bounds.

Reviewer 2: ```suggestion
When the control's `Anchor` property is set, the anchors (that is, left, top, right, bottom values) are computed and stored in an internal struct `AnchorInfo`. The only time the anchor values can be negative is when the control is placed/position outside/overlapped with the hosting control's bounds.
```

## Proposed solution

There may be cases where developers could be force-creating handles out of order, but those cases are not expected to be mainstream, and such cases will have to be manually handled by the application developer. 

The following are the events we will be using to compute anchors and replacing the current set of events mentioned in Scope section above.

- `WmCreate`,
- `OnParentChanged`,
- `SetBounds`.

> 

Reviewer 2: I would expect this section to explain the proposed solution, yet I'm not seeing it. The first sentence talks about the edge cases. Then the three events are mentioned followed by a code snippet. Followed by another sentence alluding to the past implementation and some reference to some "figure 1". And then some more code...
Where's the proposal?

I can see some bits explaining the proposal above, but they are kind of disjoined. I think the document needs to be restructured in a way that puts all proposal bits together under one section.

Author: Restructures. Take a look.

### Source snippet

```CS
private void WmCreate()
{
   ...
   DefaultLayout.UpdateAnchorInfoV2(this);
}

internal static void UpdateAnchorInfoV2(IArrangedElement element)
{
   if (!LocalAppContextSwitches.EnableAnchorLayoutV2
    || !CommonProperties.GetNeedsAnchorLayout(element))
   {
       return;
   }

>    Control control = element as Control;

Reviewer 2: ```suggestion
   Control? control = element as Control;
```

   Debug.Assert(control != null, "AnchorLayoutV2 and beyond are expected to be used only on Control type");

   if (control is null || control.Parent is null)
    {
      return;
    }

   if (!control.IsHandleCreated || !control.Parent.IsHandleCreated)
    {
      return;

>     }

Reviewer 2: Incorrect indentation

Author: Now i am copying from VS. Hopefully indentation is right. Are you using any tools to visually identify indentation here?

   ComputeAnchorInfo(IArrangedElement element)
}
```

## Simplifying Anchor calculations

> The current anchor calculation implementation was designed prior to adding high DPI support to WinForms. This proposal is replacing the current implementation with the one described in Figure 1 above. Following is the source snippet that computes the anchors.

Reviewer 2: ```suggestion
The original anchor calculation implementation was designed prior to adding high DPI support to WinForms. This proposal is replacing the current implementation with the one described in Figure 1 above. Following is the source snippet that computes the anchors.
```

Reviewer 2: I don't think the second sentence is correct - we're not replacing the original implementation; we're providing an alternative mechanism. And the original implementation can't be re-enabled with the feature switch.

Author: I did not get it. we are replacing given its default in .NET 8.0 and beyond. Developer can go back to original with switch.

```CS
private static void ComputeAnchorInfo(IArrangedElement element)
 {
    AnchorInfo? anchorInfo = GetAnchorInfo(element);
   if (anchorInfo is null)
    {
      anchorInfo = new();
      SetAnchorInfo(element, anchorInfo);
    }

   Rectangle displayRect = element.Container.DisplayRectangle;
   Rectangle elementBounds = element.Bounds;

   int x = elementBounds.X;
   int y = elementBounds.Y;

   anchorInfo.Left = x;
   anchorInfo.Top = y;

   anchorInfo.Right = displayRect.Width - (x + elementBounds.Width);
   anchorInfo.Bottom = displayRect.Height - (y + elementBounds.Height);

>  }

Reviewer 2: Incorrect indentation

# Anchor layout changes in .NET 8.0

## Description

We have multiple [issues](https://github.com/dotnet/winforms/issues?q=is%3Aissue+is%3Aopen+anchor+label%3A%22area%3A+anchor%2Fscaling%22) reported in WinForms around the anchor layout being problematic on higher DPI scale monitors, irrespective of application DPI mode. This document outlines the changes being made in .NET 8.0 to address these issues, which is aligned with the larger goal to support all supported application DPI modes in WinForms.

## Problem in Scope

> The position of an anchored control with respect to its parent should be able to determined at design time and should only need to be changed  if there were explicit changes in the control's bounds or when the control is scaled in response to a DPI changed event. Bounds changes as result of the parent control's bounds changing shouldn't alter a control's relative position in the parent's rectangle. However, currently the layout engine computes the anchored control's position every time there are changes to the control's bounds or the control's location-related properties.

Reviewer: ```suggestion
The position of an anchored control with respect to its parent should be able to be determined at design time and should only need to be changed if there were explicit changes in the control's bounds or when the control is scaled in response to a DPI changed event. Bounds changes that happen as result of the parent control's bounds changing shouldn't alter a control's relative position in the parent's rectangle. However, currently the layout engine computes the anchored control's relative position every time there are changes to the control's bounds or the control's location-related properties.
```

> For example, the following simple code snippet copied from an `InitializeComponent` method demonstrates the impact of the "over-eager" computations. The number of unnecessaryy computations can increase quite dramatically for nested layouts.

Reviewer: ```suggestion
For example, the following simple code snippet copied from a Control's `InitializeComponent` method demonstrates the impact of the "over-eager" computations. The number of unnecessary computations can increase quite dramatically for nested layouts.
```

```CS

> // The following line triggers layout to compute anchors with default button size and without parent.

Reviewer: ```suggestion
// The following line triggers layout to compute anchors with default button size and without a parent control.
```

this.button7.Anchor = ((System.Windows.Forms.AnchorStyles)((System.Windows.Forms.AnchorStyles.Left | System.Windows.Forms.AnchorStyles.Right)));

> // The following line triggers layout again to compute anchors with default button size with new location but still without parent.

Reviewer: ```suggestion
// The following line triggers layout again to compute anchors with default button size with new location but still without a parent control.
```

this.button7.Location = new System.Drawing.Point(9, 47);

> // The following line triggers layout once again to compute anchors with button's new size but still without parent.

Reviewer: ```suggestion
// The following line triggers layout once again to compute anchors with the button's new size but still without a parent.
```

this.button7.Size = new System.Drawing.Size(134, 23);

> // The following line triggers layout to compute anchors with button's current size and the default size for parent.

Reviewer: ```suggestion
// The following line triggers layout to compute anchors with the button's current size and the default size for the parent control.
```

this.Controls.Add(button7);

> // The following line triggers layout to compute anchors with button's current size and new size for parent.

Reviewer: ```suggestion
// The following line triggers layout to compute anchors with the button's current size and a new size for the parent control.
```

> // These sizes still may be changed for the DPI on the monitor depending on application's DPI mode

Reviewer: ```suggestion
// These sizes still may be changed for the current DPI depending on the application's DPI mode.
```

this.Size = new System.Drawing.Size(828, 146);

> // The following line triggers layout to compute anchors with button's

Reviewer: ```suggestion
// The following line triggers layout to compute anchors with the button's current size and a new size for the parent control.
```

current size and new size for parent.

> // DPI on the monitor still not applied.

Reviewer: ```suggestion
// The current DPI is still not applied.
```

this.ResumeLayout(false)
```
The above snippet does not represent the complete set of instances where anchor computations are unnecessary and may hold invalid anchor values. It gets even more complicated when nested UserControls are involved.

## Known issues

Customer reported issues can be largely groupped as follows.

### Missing controls

> Calculation of control's anchors with default sizes may result in negative anchors and, thus, result in invalid location for the anchored controls.

Reviewer: ```suggestion
Calculation of control's anchors with default sizes may result in negative anchors and, thus, result in an invalid location for the anchored controls.
```

![MissingControls](../images/AnchorLayoutKnownIssue_MissingControl.png)

### Overlapped Controls

> Related to the above reason, if the parent control is scaled to match the current monitor's DPI, and the control's anchor calculations happen out of sync with this scaling, controls may overlap:

Reviewer: ```suggestion
Related to the "Missing controls" issue above, if the parent control is scaled to match the current monitor's DPI, and the control's anchor calculations happen out of sync with this scaling, controls may overlap:
```

![OverlappedControls](../images/AnchorLayoutKnownIssue_OverlappedControl.png)

## Anchor Calculations

The following image illustrates anchors calculation with respect to a control's parent's display rectangle. The "red" rectangle is the parent's display rectangle, and the "green" rectangle is the anchored control's (button) bounds.
- `Left` arrow indicates the X coordinate of the button location with respect to parents display rectangle.

> - `Top` arrow indicates the Y coordinate of the button location with respect to parents display rectangle.

Reviewer: ```suggestion
- `Top` arrow indicates the Y coordinate of the button location with respect to the parent's display rectangle.
```

> - `Right` arrow indicates the distance from right edge of parent's rectangle where button is placed.

Reviewer: ```suggestion
- `Right` arrow indicates the distance from right edge of the parent's rectangle where the button is placed.
```

> - `Right` arrow indicates the distance from bottom edge of parent's rectangle where button is placed.

Reviewer: ```suggestion
- `Right` arrow indicates the distance from the bottom edge of the parent's rectangle where the button is placed.
```
